File: src/main/java/com/example/tutorial/dsa/basics/math/FizzBuzzBatchProcessor.java
package com.example.tutorial.dsa.basics.math;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

/**
 * FizzBuzzBatchProcessor
 * ----------------------------------
 * <p>This program demonstrates conditional logic and modulo operations.
 * The core problem solved here is FizzBuzz (LeetCode #412).
 *
 * <p><b>Problem Statement:</b>
 * Given an integer n, return a string array where:
 * - answer[i] == "FizzBuzz" if i is divisible by 3 and 5
 * - answer[i] == "Fizz" if i is divisible by 3
 * - answer[i] == "Buzz" if i is divisible by 5
 * - answer[i] == i (as string) otherwise
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Batch job categorization - route every 3rd offer to Marketing, every 5th to Finance</li>
 *   <li>Load balancing - distribute requests across different processing queues</li>
 *   <li>Scheduled tasks - execute different actions based on batch number</li>
 * </ul>
 *
 * <p><b>Examples:</b>
 * <ul>
 *   <li>Input: n = 3 → Output: ["1","2","Fizz"]</li>
 *   <li>Input: n = 5 → Output: ["1","2","Fizz","4","Buzz"]</li>
 *   <li>Input: n = 15 → Output: [...,"FizzBuzz"]</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> LinkedIn
 *
 * @see <a href="https://leetcode.com/problems/fizz-buzz/">LeetCode 412 - Fizz Buzz</a>
 */
@Component
public class FizzBuzzBatchProcessor implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(FizzBuzzBatchProcessor.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== FizzBuzzBatchProcessor: Conditional Logic Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate FizzBuzz with offer batch processing simulation
    System.out.println("--- Batch Processing Simulation (15 batches) ---\n");
    List<String> results = fizzBuzz(15);

    for (int i = 0; i < results.size(); i++) {
      String result = results.get(i);
      String action = switch (result) {
        case "Fizz" -> "→ Route to Marketing Team";
        case "Buzz" -> "→ Route to Finance Team";
        case "FizzBuzz" -> "→ Route to Both Teams";
        default -> "→ Process normally";
      };
      System.out.printf("Batch %2d: %-8s %s%n", i + 1, result, action);
    }
  }

  /**
   * Generates FizzBuzz sequence from 1 to n.
   *
   * <p><b>LOGIC:</b>
   * <ol>
   *   <li>Check divisibility by both 3 AND 5 first (order matters!)</li>
   *   <li>Then check divisibility by 3</li>
   *   <li>Then check divisibility by 5</li>
   *   <li>Otherwise, use the number itself</li>
   * </ol>
   *
   * <p><b>Example Walkthrough:</b>
   * <pre>
   * n = 15:
   *
   * i=1:  1%3≠0, 1%5≠0  → "1"
   * i=2:  2%3≠0, 2%5≠0  → "2"
   * i=3:  3%3=0         → "Fizz"
   * i=5:  5%5=0         → "Buzz"
   * i=15: 15%3=0, 15%5=0 → "FizzBuzz"
   * </pre>
   *
   * <p><b>Time Complexity: O(n)</b>
   * <br>We iterate from 1 to n once.
   * <br><i>Like counting items one by one - more items = more time.</i>
   *
   * <p><b>Space Complexity: O(n)</b>
   * <br>We store n strings in the result list.
   * <br><i>Like writing n labels - more batches = more labels.</i>
   *
   * @param n the number of elements to generate
   * @return list of FizzBuzz strings
   */
  public static List<String> fizzBuzz(int n) {
    List<String> result = new ArrayList<>();

    for (int i = 1; i <= n; i++) {
      // Check divisibility using modulo operator (%).
      // If remainder is 0, the number is divisible.
      boolean divisibleBy3 = (i % 3 == 0);
      boolean divisibleBy5 = (i % 5 == 0);

      // Check both conditions first (order matters!).
      if (divisibleBy3 && divisibleBy5) {
        result.add("FizzBuzz");
      } else if (divisibleBy3) {
        result.add("Fizz");
      } else if (divisibleBy5) {
        result.add("Buzz");
      } else {
        result.add(String.valueOf(i));
      }
    }

    return result;
  }
}

File: src/main/python/com/example/tutorial/dsa/basics/math/fizz_buzz_batch_processor.py
"""
FizzBuzzBatchProcessor
----------------------------------
This program demonstrates conditional logic and modulo operations.
The core problem solved here is FizzBuzz (LeetCode #412).

Problem Statement:
    Given an integer n, return a string array where:
    - answer[i] == "FizzBuzz" if i is divisible by 3 and 5
    - answer[i] == "Fizz" if i is divisible by 3
    - answer[i] == "Buzz" if i is divisible by 5
    - answer[i] == i (as string) otherwise

Real UseCase:
    In a credit card offers system:
    - Batch job categorization - route every 3rd offer to Marketing, every 5th to Finance

Examples:
    - Input: n = 3 -> Output: ["1","2","Fizz"]
    - Input: n = 15 -> Output: [...,"FizzBuzz"]

Company Tags: LinkedIn

See: https://leetcode.com/problems/fizz-buzz/
"""

import sys
import os
from typing import List

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


def fizz_buzz(n: int) -> List[str]:
    """
    Generates FizzBuzz sequence from 1 to n.

    LOGIC:
        1. Check divisibility by both 3 AND 5 first (order matters!)
        2. Then check divisibility by 3
        3. Then check divisibility by 5
        4. Otherwise, use the number itself

    Example Walkthrough:
        n = 15:

        i=1:  1%3!=0, 1%5!=0  -> "1"
        i=3:  3%3=0           -> "Fizz"
        i=5:  5%5=0           -> "Buzz"
        i=15: 15%3=0, 15%5=0  -> "FizzBuzz"

    Time Complexity: O(n)
        We iterate from 1 to n once.
        Like counting items one by one - more items = more time.

    Space Complexity: O(n)
        We store n strings in the result list.
        Like writing n labels - more batches = more labels.

    Args:
        n: The number of elements to generate.

    Returns:
        List of FizzBuzz strings.
    """
    result = []

    for i in range(1, n + 1):
        # Check divisibility using modulo operator (%).
        # If remainder is 0, the number is divisible.
        divisible_by_3 = (i % 3 == 0)
        divisible_by_5 = (i % 5 == 0)

        # Check both conditions first (order matters!).
        if divisible_by_3 and divisible_by_5:
            result.append("FizzBuzz")
        elif divisible_by_3:
            result.append("Fizz")
        elif divisible_by_5:
            result.append("Buzz")
        else:
            result.append(str(i))

    return result


def main():
    """Main function to demonstrate the FizzBuzzBatchProcessor."""
    print("=== FizzBuzzBatchProcessor: Conditional Logic Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate FizzBuzz with offer batch processing simulation
    print("--- Batch Processing Simulation (15 batches) ---\n")
    results = fizz_buzz(15)

    for i, result in enumerate(results):
        if result == "Fizz":
            action = "-> Route to Marketing Team"
        elif result == "Buzz":
            action = "-> Route to Finance Team"
        elif result == "FizzBuzz":
            action = "-> Route to Both Teams"
        else:
            action = "-> Process normally"
        print(f"Batch {i + 1:2d}: {result:8s} {action}")


if __name__ == "__main__":
    main()
